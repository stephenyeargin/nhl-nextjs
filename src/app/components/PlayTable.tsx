'use client';

import React from 'react';
import Image from 'next/image';
import SirenOnSVG from '@/app/assets/siren-on-solid.svg';
import { GAME_EVENTS } from '@/app/utils/constants';
import { formatPeriodLabel } from '@/app/utils/formatters';

interface PlayTableProps {
  plays: any[];
  activePeriod: number | null;
  renderPlayByPlayEvent: (_play: any) => React.ReactNode;
  renderTeamLogo: (_teamId?: number | string) => React.ReactNode;
}

const PlayTable: React.FC<PlayTableProps> = ({
  plays,
  activePeriod,
  renderPlayByPlayEvent,
  renderTeamLogo,
}) => {
  return (
    <div className="overflow-x-auto">
      <table className="text-xs min-w-full table-auto">
        <thead>
          <tr className="hidden">
            <th className="p-2 text-center">Time</th>
            <th className="p-2 text-center">Event Type</th>
            <th className="p-2 text-center">Details</th>
          </tr>
        </thead>
        <tbody>
          {(!plays || plays.length === 0) && (
            <tr>
              <td colSpan={3} className="p-2 text-center">
                No matching plays.
              </td>
            </tr>
          )}
          {plays &&
            plays.map((play: any, i: number) => (
              <tr key={play.eventId} className={i % 2 === 0 ? 'bg-slate-500/10' : ''}>
                <td className="p-2 text-xs text-center">
                  <div className="mt-1">
                    <span className="p-1 mx-auto font-bold border rounded">
                      {play.timeRemaining}
                    </span>
                  </div>
                  {activePeriod === 0 && (
                    <div className="p-2">{formatPeriodLabel(play.periodDescriptor, true)}</div>
                  )}
                </td>
                <td className="p-2 flex flex-wrap gap-2 items-center">
                  <div className="w-10 h-10">
                    {play.details?.eventOwnerTeamId &&
                      renderTeamLogo(play.details?.eventOwnerTeamId)}
                  </div>
                  <div>
                    <span className="hidden lg:block p-1 border rounded font-bold text-xs uppercase">
                      {(GAME_EVENTS as Record<string, string>)[play.typeDescKey]}
                    </span>
                    {play.typeDescKey === 'goal' && (
                      <div>
                        <Image
                          src={SirenOnSVG}
                          className="p-2 w-12 h-12 animate-pulse mx-auto"
                          width={200}
                          height={200}
                          alt="Goal"
                        />
                      </div>
                    )}
                  </div>
                </td>
                <td className="p-2 text-sm">{renderPlayByPlayEvent(play)}</td>
              </tr>
            ))}
        </tbody>
      </table>
    </div>
  );
};

export default PlayTable;
