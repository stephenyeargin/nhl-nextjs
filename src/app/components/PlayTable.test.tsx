import React from 'react';
import { render, screen } from '@testing-library/react';
import PlayTable from './PlayTable';

const plays = [
  {
    eventId: 1,
    timeRemaining: '12:34',
    periodDescriptor: { number: 1 },
    details: { eventOwnerTeamId: 1 },
    typeDescKey: 'goal',
  },
  {
    eventId: 2,
    timeRemaining: '05:00',
    periodDescriptor: { number: 1 },
    details: { eventOwnerTeamId: 2 },
    typeDescKey: 'hit',
  },
];

describe('PlayTable', () => {
  it('renders rows and delegates details and logos', () => {
    const { asFragment } = render(
      <PlayTable
        plays={plays as any}
        activePeriod={0}
        renderPlayByPlayEvent={(play) => <div data-testid={`details-${play.eventId}`}>D</div>}
        renderTeamLogo={(teamId?: number | string) => <div data-testid={`logo-${teamId}`} />}
      />
    );
    expect(screen.getByText('12:34')).toBeInTheDocument();
    expect(screen.getByTestId('details-1')).toBeInTheDocument();
    expect(screen.getByTestId('logo-1')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders empty state when no plays', () => {
    const { asFragment } = render(
      <PlayTable
        plays={[]}
        activePeriod={1}
        renderPlayByPlayEvent={() => null}
        renderTeamLogo={() => null}
      />
    );
    expect(screen.getByText('No matching plays.')).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
