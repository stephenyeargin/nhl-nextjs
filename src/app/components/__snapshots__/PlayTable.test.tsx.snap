// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayTable renders empty state when no plays 1`] = `
<DocumentFragment>
  <div
    class="overflow-x-auto"
  >
    <table
      class="text-xs min-w-full table-auto"
    >
      <thead>
        <tr
          class="hidden"
        >
          <th
            class="p-2 text-center"
          >
            Time
          </th>
          <th
            class="p-2 text-center"
          >
            Event Type
          </th>
          <th
            class="p-2 text-center"
          >
            Details
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="p-2 text-center"
            colspan="3"
          >
            No matching plays.
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`PlayTable renders rows and delegates details and logos 1`] = `
<DocumentFragment>
  <div
    class="overflow-x-auto"
  >
    <table
      class="text-xs min-w-full table-auto"
    >
      <thead>
        <tr
          class="hidden"
        >
          <th
            class="p-2 text-center"
          >
            Time
          </th>
          <th
            class="p-2 text-center"
          >
            Event Type
          </th>
          <th
            class="p-2 text-center"
          >
            Details
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="bg-slate-500/10"
        >
          <td
            class="p-2 text-xs text-center"
          >
            <div
              class="mt-1"
            >
              <span
                class="p-1 mx-auto font-bold border rounded"
              >
                12:34
              </span>
            </div>
            <div
              class="p-2"
            >
              1st Period
            </div>
          </td>
          <td
            class="p-2 flex flex-wrap gap-2 items-center"
          >
            <div
              class="w-10 h-10"
            >
              <div
                data-testid="logo-1"
              />
            </div>
            <div>
              <span
                class="hidden lg:block p-1 border rounded font-bold text-xs uppercase"
              >
                Goal
              </span>
              <div>
                <img
                  alt="Goal"
                  class="p-2 w-12 h-12 animate-pulse mx-auto"
                  data-nimg="1"
                  decoding="async"
                  height="200"
                  loading="lazy"
                  src="/_next/image?url=%2Fimg.jpg&w=640&q=75"
                  srcset="/_next/image?url=%2Fimg.jpg&w=256&q=75 1x, /_next/image?url=%2Fimg.jpg&w=640&q=75 2x"
                  style="color: transparent;"
                  width="200"
                />
              </div>
            </div>
          </td>
          <td
            class="p-2 text-sm"
          >
            <div
              data-testid="details-1"
            >
              D
            </div>
          </td>
        </tr>
        <tr
          class=""
        >
          <td
            class="p-2 text-xs text-center"
          >
            <div
              class="mt-1"
            >
              <span
                class="p-1 mx-auto font-bold border rounded"
              >
                05:00
              </span>
            </div>
            <div
              class="p-2"
            >
              1st Period
            </div>
          </td>
          <td
            class="p-2 flex flex-wrap gap-2 items-center"
          >
            <div
              class="w-10 h-10"
            >
              <div
                data-testid="logo-2"
              />
            </div>
            <div>
              <span
                class="hidden lg:block p-1 border rounded font-bold text-xs uppercase"
              >
                Hit
              </span>
            </div>
          </td>
          <td
            class="p-2 text-sm"
          >
            <div
              data-testid="details-2"
            >
              D
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
